### PR TITLE
feat(moe_bgmv): direct decode fast-path for shrink (1.3-3.8x vs sliced kernel)

### PR DESCRIPTION
**Pull request by aws-jiadingg**
## 📌 Description

Adds a fused decode fast-path to `moe_bgmv_shrink_sliced`. In the decode path
(`num_pairs <= decode_threshold`, i.e. small batch), the existing cp.async-pipelined
sliced kernel is launch-starved: its grid is only
`ceil(num_pairs/PPB) × ceil(feat_out/RANK_TILE) × num_slices` blocks, so just a handful
of SMs are active, and the shared-memory staging / async-pipeline overhead is pure cost
because there is no inter-pair reuse to amortize it at small batch.

The new `moe_bgmv_shrink_fused_kernel` maps **one CTA per (pair, rank, slice)** output
element. Each block strides the full `feat_in` contraction with 128-bit vectorized loads
(`vec_t`), reduces with a warp shuffle + small block reduction, and fuses the
`scale + cast` into the single accumulating write. This raises resident blocks ~`feat_out`-fold
and removes the staging overhead.

## 🔍 Related Issues

N/A

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [x] All tests are passing (`unittest`, etc.).

Validated via FlashInfer's own `tests/moe/test_bgmv_moe.py` (JIT recompile of the edited source):

```
tests/moe/test_bgmv_moe.py ............ 147 passed
```

- `TestBgmvMoeShrink` 48/48, `TestBgmvMoeEndToEnd` + edge cases 27/27, full suite 147/147.
- Batch sizes give `num_pairs ∈ {2, 8, 64}`, exercising both the fused path (≤32) and the
  pipelined path (64).
- **Added `test_shrink_correctness_multislice`** (`num_slices=2`, gate_up packing, bf16+fp16)
  to cover the kernel's `blockIdx.z` slice path, which the existing `num_slices=1` tests do
  not exercise — 24/24 pass, `max_diff = 0`.
- `compute-sanitizer` clean on the fused kernel: **memcheck 0 errors, racecheck 0 hazards,
  synccheck 0 errors** (decode / multi-slice / `lora_id=-1` skip / out-of-bounds-token branches).

## Reviewer Notes

**What the speedup compares.** All numbers are **this PR's `moe_bgmv_shrink_fused_kernel`
vs. the existing `moe_bgmv_shrink_sliced_kernel` decode instantiation** (the current
upstream baseline: `RANK_TILE=8, PAIRS_PER_BLOCK=4, NUM_STAGES=3` — the cp.async-pipelined
kernel selected today when `num_pairs <= decode_threshold`). Both kernels were compiled from
the same tree and run through the same `bgmv_moe_shrink` API on the same H200 (SM90), at the
same shape (`feat_in=3072, feat_out=32, bf16, num_pairs=8`), timed back-to-back. So this is a
like-for-like replacement of one decode kernel by another.

**SASS comparison insights** (control codes decoded with the SM90 layout — `hi64[44:41]`
stall, `[48:46]` write-barrier, `[57:52]` wait mask; same disassembled `.cuda.o`):

| | baseline `..._sliced_kernel` (decode) | this PR `..._fused_kernel` |
|---|---|---|
| static instructions | 4008 | **184** (21.8× fewer) |
| total static stall cycles | 11 798 | 410 |
| `LDGSTS` async-copy ops | 12 | **0** |
| smem staging (`STS.128`/`LDS.128`) | present (pipeline round-trip) | **none** |

### Benchmarking results on H200

Kernel-level evidence at `feat_in=3072, feat_out=32, bf16`, decode (`num_pairs=8`), on the
production build path (`bgmv_moe_bf16_bf16_bf16`), H200 SM90 — i.e. NOT the Python/API
dispatch floor:

| metric | baseline (pipelined) | this PR (direct) |  |
|---|---|---|---|
| **kernel duration** (ncu `gpu__time_duration`) | **34.5 µs** | **5.95 µs** | **≈5.8× faster kernel** |
| blocks launched | 8 | 256 | 32× parallelism |
| occupancy limit (smem) | **1 block/SM** | 28 blocks/SM | staging cap removed |
| SM throughput | 0.87 % | 5.12 % | |
| `LDGSTS` async-copy ops (SASS) | 12 | **0** | |
| static instructions (SASS) | 4008 | **184** | 21.8× fewer |

The mechanism is structural (launch starvation + cp.async-staging removal) and reproducible
under ncu, not a fragile micro-optimization. The prefill kernel's SASS is byte-for-byte
identical before/after this change.

### Performance ( `bgmv_moe_shrink` API, H200 SM90)

CUDA-event timed, 30 warmup + 200 iters × 5 reps, median; baseline vs this PR, same GPU:

| hidden | rank | tokens | baseline µs | this PR µs | speedup |
|-------:|-----:|-------:|------------:|-----------:|--------:|
|   768  |  16  |   4    |    21.4     |    16.0    | **1.34×** |
|  2048  |  16  |  16    |    30.5     |    15.9    | **1.92×** |
|  3072  |  32  |   4    |    38.4     |    16.0    | **2.40×** |
|  4096  |  32  |  16    |    46.3     |    16.9    | **2.73×** |
|  5888  |  32  |   4    |    61.6     |    16.0    | **3.84×** |

Win grows with hidden size. Prefill control (tokens=256, gated → pipelined kernel): within
noise, no regression. (The ~16 µs API floor is Python dispatch; the kernel speedup is larger,
as the ncu numbers above show.)

> The core CUDA optimization techniques in this patch were discovered by IFCO, an autonomous
> CUDA optimization agent built by AWS; the kernel was then re-derived against the production `w_ptr` table
> interface.

| hot-path global load | mixed | single rolled **`LDG.E.128`** (0x800-B grid stride) |
| scalar `LDG.E.U16` fallback | — | **0** (`if constexpr`-compiled away) |

- The baseline spends most of its instruction stream staging `X`/`W` tiles into shared
  memory via `LDGSTS` and reading them back (`LDS.128`) to feed the pipeline. At decode there
  is no inter-pair reuse to amortize that round-trip, so it is pure latency. The fused kernel
  reads global memory straight into registers with 128-bit `LDG.E.128` and reduces with
  `SHFL` — eliminating all `LDGSTS`/`STS`/`LDS` and 21.8× of the static instructions.
- The disassembly confirms the contraction is **fully 128-bit vectorized** (one `LDG.E.128`
  loop, no scalar `LDG.E.U16` path), and ncu reports 9.78 sectors/request — consistent with
  128-bit loads, not a scalar bf16 contraction.

## Summary by CodeRabbit

* **New Features**
  * Added an optimized decode fast-path for MoE BGMV shrink computations to speed up execution using vectorized operations and an early-exit dispatch when supported.

* **Tests**
  * Added a new PyTest case to validate shrink correctness for multi-slice runs (`num_slices=2`) by comparing CUDA results against the existing reference implementation.

### CODE REVIEW

**Comment by jhaotingc** [comment]
LGTM for this.
Hi @yzh119, this is a follow up PR from AWS's BGMV kernel. Would you mind reviewing this?
This adds a separate path for decode where it addressed case when low CTA is launched.

Thank you!!

File: csrc/bgmv_moe/moe_bgmv_impl.cuh
Comment (taehokim20) [inline csrc/bgmv_moe/moe_bgmv_impl.cuh:520]: Could you restore the final newline? The repo's .pre-commit-config.yaml runs end-of-file-fixer, so pre-commit run --all-files would flag both. test_bgmv_moe.py as well.

File: csrc/bgmv_moe/moe_bgmv_impl.cuh [hunk 1]
```diff
@@ -325,6 +397,20 @@ void moe_bgmv_shrink_sliced(out_T* __restrict__ Y, const in_T* __restrict__ X,
   const bool extended = (sm_major >= 9);
   const bool decode = (num_pairs <= MoeShrinkKernelConfig::decode_threshold);
 
+  // Decode fast-path: one block per (pair, rank, slice) with vectorized contraction.
+  // All compiled `wide` (feat_in) values are multiples of 8, so 128-bit loads apply.
+  if (decode) {
```
Comment (taehokim20) [inline csrc/bgmv_moe/moe_bgmv_impl.cuh]: The baseline gates the decode tuning behind `extended = (sm_major >= 9);`. The new fast-path is dispatched on `if (decode)` unconditionally, so on sm_80 and below the decode path now also routes to the direct kernel instead of the pipelined one. Is it intended?

File: csrc/bgmv_moe/moe_bgmv_impl.cuh
Comment (aws-jiadingg) [reply csrc/bgmv_moe/moe_bgmv_impl.cuh:520]: Restored — both `csrc/bgmv_moe/moe_bgmv_impl.cuh` and `tests/moe/test_bgmv_moe.py` now end with a single trailing newline (pre-commit run --all-files passes clean, no end-of-file-fixer diff). Thanks for the catch.

File: csrc/bgmv_moe/moe_bgmv_impl.cuh (hunk 1 above)
Comment (aws-jiadingg) [reply csrc/bgmv_moe/moe_bgmv_impl.cuh]: Good catch — not intended. The fast-path is only measured/tuned on Hopper+, so it shouldn't silently reroute pre-Hopper. Fixed in a36b800c: the dispatch is now if (extended && decode), matching the existing extended gate on the pairs_per_block/num_stages decode tuning right below it. On sm_80 and below the decode path keeps the proven pipelined sliced kernel. Thanks!

**Comment by jhaotingc** [comment]
Hi @taehokim20 can you leave a approve to this PR (not sure if you can or not) or leave a LGTM indicating AWS reviewed. Thanks!

**Comment by taehokim20** [comment]
LGTM. However, I don't think I have the authority to approve this PR.

**Review by aleozlx — APPROVED** [review]
